specs/core/1_new_shards.md

```diff
@@ -0,0 +1,440 @@
+            all_pubkeys.append(state.validators[participant].pubkey)
+            # Note: only 2N distinct message hashes
+            all_message_hashes.append(AttestationCustodyBitWrapper(hash_tree_root(indexed_attestation.data), i, bit))
+        
+    return bls_verify_multiple(
+        pubkeys=all_pubkeys,
+        message_hashes=all_message_hashes,
+        signature=indexed_attestation.signature,
+        domain=get_domain(state, DOMAIN_BEACON_ATTESTER, indexed_attestation.data.target.epoch),
+    )
+```
+
+## Beacon Chain Changes
+
+### New state variables
+
+```python
+    shard_states: Vector[ShardState, MAX_SHARDS]
+    online_countdown: Bytes[VALIDATOR_REGISTRY_LIMIT]
+    current_light_committee: CompactCommittee
+    next_light_committee: CompactCommittee
+```
+
+### New block data structures
+
+```python
+    shard_transitions: Vector[ShardTransition, MAX_SHARDS]
+    light_client_signature_bitfield: Bitlist[LIGHT_CLIENT_COMMITTEE_SIZE]
+    light_client_signature: BLSSignature
+```
+
+### Attestation processing
+
+```python
+def process_attestation(state: BeaconState, attestation: Attestation) -> None:
+    data = attestation.data
+    assert data.index < ACTIVE_SHARDS
+    shard = (data.index + get_start_shard(state, data.slot)) % ACTIVE_SHARDS
+    proposer_index=get_beacon_proposer_index(state)
+
+    # Signature check
+    committee = get_crosslink_committee(state, get_current_epoch(state), shard)
+    for bits in attestation.custody_bits + [attestation.aggregation_bits]:
+        assert len(bits) == len(committee)
+    # Check signature
+    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
+    # Get attesting indices
+    attesting_indices = get_attesting_indices(state, attestation.data, attestation.aggregation_bits)
+
+    # Prepare pending attestation object
+    pending_attestation = PendingAttestation(
+        slot=data.slot,
+        shard=shard,
+        aggregation_bits=attestation.aggregation_bits,
+        inclusion_delay=state.slot - data.slot,
+        crosslink_success=False,
+        proposer_index=proposer_index
+    )
+    
+    # Type 1: on-time attestations
+    if data.custody_bits != []:
+        # Correct slot
+        assert data.slot == state.slot
+        # Slot the attestation starts counting from
+        start_slot = state.shard_next_slots[shard]
+        # Correct data root count
+        offset_slots = [start_slot + x for x in SHARD_BLOCK_OFFSETS if start_slot + x < state.slot]
+        assert len(attestation.custody_bits) == len(offset_slots)
+        # Correct parent block root
+        assert data.beacon_block_root == get_block_root_at_slot(state, state.slot - 1)
+        # Apply
+        online_indices = get_online_indices(state)
+        if get_total_balance(state, online_indices.intersection(attesting_indices)) * 3 >= get_total_balance(state, online_indices) * 2:
+            # Check correct formatting of shard transition data
```

Comment:
Need to consider what happens if a fully duplicate attestation (or a not fully duplicate attestation, but both have 2/3 voting on a crosslink) is included. This would create two `PendingAttestations` with `crosslink_success` as `True`

Reply:
Depending on how rewards are done, it is likely fine. Just need to get the union of the successful participants at `slot/index`.